File: enterprise-v2/backend/app/data/branch_identity.py
"""Canonical identity for line-of-business display names."""

from __future__ import annotations

import re
import unicodedata
from collections.abc import Iterable


def branch_identity_key(value: object) -> str:
    text = unicodedata.normalize("NFKD", str(value or ""))
    text = "".join(char for char in text if not unicodedata.combining(char))
    text = text.replace("ı", "i").lower()
    return re.sub(r"\s+", " ", text).strip()


def same_branch_name(left: object, right: object) -> bool:
    return branch_identity_key(left) == branch_identity_key(right)


def clean_branch_display_name(value: object) -> str:
    return re.sub(r"\s+", " ", str(value or "")).strip()


def unique_branch_names(values: Iterable[object]) -> list[str]:
    names: dict[str, str] = {}
    for value in values:
        display = clean_branch_display_name(value)
        key = branch_identity_key(display)
        if key and key not in names:
            names[key] = display
    return list(names.values())

File: enterprise-v2/frontend/src/lib/branch-identity.ts
/**
 * Branch names are display labels, not stable identifiers. All joins use this
 * canonical key so casing, Turkish dotted/dotless I, accents and spacing do not
 * split one logical line of business into multiple branches.
 */
export function branchIdentityKey(value: string | null | undefined): string {
  return String(value ?? "")
    .normalize("NFKD")
    .replace(/\p{M}/gu, "")
    .replace(/ı/g, "i")
    .toLowerCase()
    .trim()
    .replace(/\s+/g, " ");
}

export function sameBranchName(
  left: string | null | undefined,
  right: string | null | undefined,
): boolean {
  return branchIdentityKey(left) === branchIdentityKey(right);
}

export function cleanBranchDisplayName(value: string | null | undefined): string {
  return String(value ?? "").trim().replace(/\s+/g, " ");
}

/** Deduplicate by identity while preserving the first source spelling. */
export function uniqueBranchNames(
  values: Iterable<string | null | undefined>,
): string[] {
  const names = new Map<string, string>();
  for (const value of values) {
    const display = cleanBranchDisplayName(value);
    const key = branchIdentityKey(display);
    if (key && !names.has(key)) names.set(key, display);
  }
  return [...names.values()];
}
